core/shared/src/main/scala/scalaz/zio/Semaphore.scala

```diff
@@ -22,26 +24,37 @@ final class Semaphore private (private val state: Ref[State]) extends Serializab
   final def release: IO[Nothing, Unit] = releaseN(1)
 
   final def withPermit[E, A](task: IO[E, A]): IO[E, A] =
-    IO.bracket[E, Unit, A](acquire)(_ => release)(_ => task)
-
-  final def acquireN(requested: Long): IO[Nothing, Unit] = {
-    val acquire: (Promise[Nothing, Unit], State) => (IO[Nothing, Boolean], State) = {
-      case (p, Right(n)) if n >= requested => p.complete(()) -> Right(n - requested)
-      case (p, Right(n))                   => IO.now(false)  -> Left(IQueue(p -> (requested - n)))
-      case (p, Left(q))                    => IO.now(false)  -> Left(q.enqueue(p -> requested))
-    }
+    IO.bracket0[E, AcquireTasks, A](prepare(1L))(cleanup) { case (acquire, _) => acquire *> task }
+
+  final def acquireN(requested: Long): IO[Nothing, Unit] =
+    assertNonNegative(requested) *> IO.bracket0[Nothing, AcquireTasks, Unit](prepare(requested))(cleanup)(_._1)
+
+  final private def prepare(n: Long): IO[Nothing, AcquireTasks] = {
```

Comment:
All credit should go to @mpilquist for finding out the solution in https://github.com/typelevel/cats-effect/pull/403, so let's mention that in a comment.